### DISCUSSION TITLE
How does ghostty get the system light/dark setting on Linux?

### DISCUSSION THREAD

**Discussion by hankertrix** [ghostty-org/ghostty#2980] [Q&A] [answered]
I run Ghostty in a tiling window manager and don't use a desktop environment, so I'm trying to find the option to set the system light/dark setting.

I tried

```sh
gsettings set org.gnome.desktop.interface color-scheme 'prefers-dark'
```

But it doesn't seem to make a difference, as Ghostty will always default to the light colour scheme regardless of what that value is.

**Comment by jcollie** [comment]
You'll need to double-check as I'm on mobile now but you need to set `window-theme=system` in your config.

**Reply by hankertrix** [reply]
I have done that, but nothing changed. Previously it was set to `ghostty`, so it should have been the same as having it set to `system`.

**Reply by jcollie** [reply]
Ghostty uses a DBus protocol to get the system theme so if your desktop environment does not implement that you may be out of luck (and yes, a tiling WM counts as a desktop environment even if its not a traditional one).

**Reply by hankertrix** [reply]
Ah I see. Do you know the protocol name, or the dbus command to get the value from the protocol?

I'm using i3wm, not sure if it implements such a protocol.

**Reply by jcollie** [reply, marked as answer]
Ghostty uses the [Freedesktop Settings Portal](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Settings.html) to listen for theme changes.

https://github.com/ghostty-org/ghostty/blob/6a1df9fb8d845c728548d4550bdbe3b14788ace7/src/apprt/gtk/App.zig#L1116-L1134

**Reply by hankertrix** [reply]
Ohhhh I see. Yeap then I'm out of luck. That's a Wayland-specific protocol that i3 will never provide since it's using X Org, but thanks for the reply!

For anyone looking to read the colour scheme setting manually using this protocol, use the command below, taken from https://github.com/pbek/QOwnNotes/issues/2525:
```sh
dbus-send --session --print-reply=literal --reply-timeout=1000 --dest=org.freedesktop.portal.Desktop /org/freedesktop/portal/desktop org.freedesktop.portal.Settings.Read string:'org.freedesktop.appearance' string:'color-scheme'
```

**Reply by jcollie** [reply]
It's definitely not Wayland-specific, it works just fine on Gnome/X11 and KDE/X11. You're right though that i3 may not provide support for that protocol.

**Reply by hankertrix** [reply]
Ahhh I see. Looks like I confused portals with Wayland protocols when it’s for Flatpaks.

Looks like I just need to install an implementation of `xdg-desktop-portal` and start the daemon to get the themes working.

**Comment by hankertrix** [comment]
For anyone trying to figure out how to get `xdg-desktop-portal` set up to have light and dark themes working, especially if you are using a tiling window manager or the like that doesn't have a default portal implementation, here's how to do so:

1. Install `xdg-desktop-portal` and a `xdg-desktop-portal-impl`. As a generic fallback implementation for most tiling window managers, you should use `xdg-desktop-portal-gtk`. If you use a desktop environment, like GNOME or KDE, then install the portal for your desktop environment, like `xdg-desktop-portal-gnome` for GNOME and `xdg-desktop-portal-kde` for KDE.
2. By default, the portal that you install will only be applied if you have the same `XDG_CURRENT_DESKTOP` value as the portal implementation that you install. For `xdg-desktop-portal-gtk`, the value is `gtk`, so if you are on `i3`, or `qtile`, or `awesomewm`, or any other tiling window manager for that matter, the portal implementation will not be used by default.
3. If you would like to have automatic switching between light and dark themes, install [`darkman`](https://gitlab.com/WhyNotHugo/darkman) and configure it by creating `~/.config/darkman/config.yaml` and entering the values below:

```
lat: ${your_latitude_as_float}
lng: ${your_longitude_as_float}
usegeoclue: false
dbusserver: true
portal: true
```

Where `${your_latitude_as_float}` and `${your_longitude_as_float}` are your location's latitude and longitude as floating point numbers respectively. Note that `portal` must be set to `true` to have Ghostty and most other terminals recognise the system theme through `darkman`'s settings portal.

4. Create the `~/.config/xdg-desktop-portal/portals.conf` file, or `~/.config/xdg-desktop-portal/${XDG_CURRENT_DESKTOP}-portals.conf`, where `${XDG_CURRENT_DESKTOP}` is the value when you run `echo $XDG_CURRENT_DESKTOP` in your shell. The first file will take effect no matter what the value of `XDG_CURRENT_DESKTOP` is, so you should create the second file if you would like to configure `xdg-desktop-portal` for only your current window manager or desktop environment and not any other.
5. Enter the values below into the file:

```
[preferred]
default=gtk
org.freedesktop.impl.portal.Settings=darkman
```

Change `gtk` to whichever `xdg-desktop-portal-impl` you choose to use, like `gnome` for GNOME and `kde` for KDE. You can delete the last line if you don't want to use `darkman` to toggle between the light and dark themes automatically.

6. Enjoy switching between your light and dark themes!
